src/backend/distributed/commands/trigger.c

```diff
@@ -203,45 +255,602 @@ get_relation_trigger_oid_compat(HeapTuple heapTuple)
+}
+
+
+/*
+ * ErrorOutForTriggerCommandIfNotCitusLocalTable is a helper function to error
+ * out for unsupported trigger commands depending on the citus table type.
+ */
+void
+ErrorOutForTriggerCommandIfNotCitusLocalTable(Oid relationId)
+{
+	if (IsCitusLocalTable(relationId))
 	{
 		return;
 	}
 
-	char *relationName = triggerRelation->relname;
+	ereport(ERROR, (errmsg("triggers are only supported for citus local tables")));
+}
+
+
+/*
+ * GetDropTriggerStmtRelation takes a DropStmt for a trigger object and returns
+ * RangeVar for the relation that owns the trigger.
+ */
+static RangeVar *
+GetDropTriggerStmtRelation(DropStmt *dropTriggerStmt)
+{
+	Assert(dropTriggerStmt->removeType == OBJECT_TRIGGER);
+
+	ErrorIfDropStmtDropsMultipleTriggers(dropTriggerStmt);
+
+	List *targetObjectList = dropTriggerStmt->objects;
+	List *triggerObjectNameList = linitial(targetObjectList);
+
+	/*
+	 * The name list that identifies the trigger to be dropped looks like:
+	 * [catalogName, schemaName, relationName, triggerName], where, the first
+	 * two elements are optional. We should take all elements except the
+	 * triggerName to create the range var object that defines the owner
+	 * relation.
+	 */
+	int relationNameListLength = list_length(triggerObjectNameList) - 1;
+	List *relationNameList = list_truncate(list_copy(triggerObjectNameList),
+										   relationNameListLength);
+
+	return makeRangeVarFromNameList(relationNameList);
+}
+
+
+/*
+ * DropTriggerEventExtendNames extends relation name and trigger name with
+ * shardId, and sets schema name in given DropStmt by recreating "objects"
+ * list.
+ */
+void
+DropTriggerEventExtendNames(DropStmt *dropTriggerStmt, char *schemaName, uint64 shardId)
+{
+	Assert(dropTriggerStmt->removeType == OBJECT_TRIGGER);
+
+	char *triggerName = NULL;
+	char *relationName = NULL;
+	ExtractDropStmtTriggerAndRelationName(dropTriggerStmt, &triggerName, &relationName);
+
+	AppendShardIdToName(&triggerName, shardId);
+	Value *triggerNameValue = makeString(triggerName);
+
+	AppendShardIdToName(&relationName, shardId);
+	Value *relationNameValue = makeString(relationName);
+
+	Value *schemaNameValue = makeString(pstrdup(schemaName));
+
+	List *shardTriggerNameList =
+		list_make3(schemaNameValue, relationNameValue, triggerNameValue);
+	dropTriggerStmt->objects = list_make1(shardTriggerNameList);
+}
+
+
+/*
+ * ExtractDropStmtTriggerAndRelationName extracts triggerName and relationName
+ * from given dropTriggerStmt if arguments are passed as non-null pointers.
+ */
+static void
+ExtractDropStmtTriggerAndRelationName(DropStmt *dropTriggerStmt, char **triggerName,
+									  char **relationName)
+{
+	ErrorIfDropStmtDropsMultipleTriggers(dropTriggerStmt);
+
+	List *targetObjectList = dropTriggerStmt->objects;
+	List *triggerObjectNameList = linitial(targetObjectList);
+	int objectNameListLength = list_length(triggerObjectNameList);
+
+	if (triggerName != NULL)
+	{
+		int triggerNameindex = objectNameListLength - 1;
```

Comment:
Can you create a utility method or macro such as:

```c
void EnsureNonnegative(int num) {
    if (num < 0 ) {
         ereport(ERROR, errmsg("%d should be a nonnegative number"));
    }
}
```
and call that with triggerNameIndex and relationNameIndex below? I just want to be on the safe side with indexes.

Alternatively we could introduce a `safe_list_nth` which checks the index before calling `list_nth`.